### PR TITLE
enhance: add functions for daemon servers for mTLS

### PR DESCRIPTION
**Pull request by g-linville**
Here is an example of what it looks like to use these functions:

```go
package main

import (
	"fmt"
	"net/http"
	"os"

	"github.com/gptscript-ai/go-gptscript/pkg/daemon"
)

func main() {
	server, err := daemon.CreateServer()
	if err != nil {
		fmt.Printf("failed to create server: %v\n", err)
		os.Exit(1)
	}

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		clientCert := r.TLS.PeerCertificates[0]
		fmt.Fprintf(w, "Hello, %s\n", clientCert.Subject.CommonName)
	})

	if err := daemon.StartServer(server); err != nil {
		fmt.Printf("failed to start server: %v\n", err)
		os.Exit(1)
	}
}
```

### CODE REVIEW

**Review by iwilltry42 — APPROVED** [review]

**Review by drpebcak — APPROVED** [review]

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+// This function should be used when creating a new server for a daemon tool with a custom ServeMux.
+// The server should then be started with the StartServer function.
+func CreateServerWithMux(mux *http.ServeMux) (*http.Server, error) {
+	tlsConfig, err := getTLSConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get TLS config: %v", err)
+	}
+
+	server := &http.Server{
+		Addr:      fmt.Sprintf("127.0.0.1:%s", os.Getenv("PORT")),
```
Comment (thedadams) [inline pkg/daemon/daemon.go:38]: We probably want to make the `port` a parameter. Otherwise, you could only ever run one daemon at a time.

**Review by thedadams — COMMENTED** [review]
I'm fine with the two functions for creating a server, but it seems strange to call `CreateServer()`, add a bunch of stuff to the default `http` server mux, and then `StartServer`.

Instead of returning an `*http.Server`, can we return our own server type that has methods like the `http.ServerMux` that just passes them onto an `*http.ServerMux` field on the server type. Something like:

```go
type Server struct {
 	 mux *http.ServeMux
 	 tlsConfig *tls.Config
 	 port string
}

func CreateServer(port string) (*Server, error) {
 	retrurn CreateServerWithMux(http.DefaultServeMux, port)
 }

func CreateServerWithMux(mux *http.ServeMux, port string) (*Server, error) {
 	tlsConfig, err := getTLSConfig()
 	if err != nil {
 		return nil, fmt.Errorf("failed to get TLS config: %v", err)
 	}

 	 server := &Server{
 	  	 tlsConfig: tlsConfig,
 	  	 port: port,
 	  	 mux: mux,
 	 }
 	return server, nil
 }

func (s *Server) HandleFunc(patter string, handler http.HandlerFunc) {
 	 s.mux.HandleFunc(pattern, handler)
}
```

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		TLSConfig: tlsConfig,
+		Handler:   mux,
+	}
+	return server, nil
+}
+
+// StartServer starts an HTTP server created by the CreateServer function.
+// This is for use with daemon tools.
+func StartServer(server *http.Server) error {
+	if err := server.ListenAndServeTLS("", ""); err != nil {
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/daemon/daemon.go]: fixed

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (g-linville) [reply pkg/daemon/daemon.go:44]: But we do only want to run one daemon at a time, at least one per process, and it is required to use the port that GPTScript chose for it. So I don't think there is a point in making this a parameter. This is just a library function to be called in each daemon tool, not something that will get used anywhere else.

**Review by thedadams — APPROVED** [review]

**Pull request merged**